frontend/src/models/propertyDefinitionsModel.ts

```diff
@@ -1,163 +1,223 @@
-                    count: state.count,
-                    results: state.results.map((p) => (property.id === p.id ? property : p)),
-                    next: state.next,
+                updatePropertyDefinition: (state, { propertyDefinition }) => ({
+                    ...state,
+                    [propertyDefinition.name]: propertyDefinition,
+                }),
+                updatePropertyDefinitions: (state, { propertyDefinitions }) => ({
+                    ...state,
+                    ...(Array.isArray(propertyDefinitions)
+                        ? Object.fromEntries(propertyDefinitions.map((p) => [p.name, p]))
+                        : propertyDefinitions),
                 }),
             },
         ],
     }),
-    listeners: ({ actions }) => ({
-        loadPropertyDefinitionsSuccess: ({ propertyStorage }) => {
-            if (propertyStorage.next) {
-                actions.loadPropertyDefinitions()
+    listeners(({ actions, values }) => ({
+        loadPropertyDefinitions: async ({ propertyKeys }) => {
+            const { propertyDefinitionStorage } = values
+
+            let mustLoad = false
+            const pendingStateUpdate: PropertyDefinitionStorage = {}
+            for (const key of propertyKeys) {
+                if (
+                    !(key in propertyDefinitionStorage) ||
+                    propertyDefinitionStorage[key] === PropertyDefinitionState.Error
+                ) {
+                    mustLoad = true
+                    pendingStateUpdate[key] = PropertyDefinitionState.Pending
+                }
             }
+
+            // nothing new to do
+            if (!mustLoad) {
+                return
+            }
+
+            // set all requested properties as `PropertyDefinitionState.Pending`
+            actions.updatePropertyDefinitions(pendingStateUpdate)
+            // run the next part of this chain
+            actions.fetchAllPendingDefinitions()
         },
-    }),
-    events: ({ actions }) => ({
-        afterMount: () => {
-            actions.loadPropertyDefinitions(true)
+        fetchAllPendingDefinitions: async (_, breakpoint) => {
+            // take 10ms to debounce property definition requests, preventing a lot of small queries
+            await breakpoint(10)
+            if (values.pendingProperties.length === 0) {
+                return
+            }
+            // take the first 50 pending properties to avoid the 4k query param length limit
+            const pending = values.pendingProperties.slice(0, 50)
+            try {
+                // set them all as PropertyDefinitionState.Loading
+                actions.updatePropertyDefinitions(
+                    Object.fromEntries(pending.map((key) => [key, PropertyDefinitionState.Loading]))
+                )
+                // and fetch them
+                const { url } = combineUrl('api/projects/@current/property_definitions/', {
```

Comment:
should the URL combination be in `api.ts` instead of building it and calling `api.get` here?

Reply:
I think it should. Moved it!